### PR TITLE
#4046 Split Shabondi boot class and definition into source and sink groups

### PR DESCRIPTION
**Pull request by vitojeng**
required by #4046 

### Checklist
- [x] [Review Contribution Guidelines](https://ohara.readthedocs.io/en/latest/contributing.html)
- [x] Add Reviewers (see [Authors](https://github.com/oharastream/ohara#ohara-team))
- [x] Add Assignees (of course, you are the assignee by default!)
- [x] DON'T set a Milestone
- [x] DON'T add any labels (PR labels will be added automatically by Jenkins)
- [x] Link to an issue(You can do so by using the required by keyword: required by #issue_number)
- [x] Add unit tests (or please explain why this PR is not worth having new tests)
- [ ] Pass QA (Sometimes you may encounter an existing flaky test which would obstruct you from getting QA passed. If the failed tests are unrelated, you can add a comment in the PR thread)

### CODE REVIEW

File: ohara-shabondi/src/main/scala/oharastream/ohara/shabondi/ShabondiDefinitions.scala [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+import scala.collection.JavaConverters._
+import scala.collection.mutable
+
+sealed abstract class ShabondiType(val name: String, val className: String)
+
+final case object ShabondiSource extends ShabondiType("source", SourceBoot.getClass.getName.replaceAll("\\$", ""))
+
+final case object ShabondiSink extends ShabondiType("sink", SinkBoot.getClass.getName.replaceAll("\\$", ""))
+
+final object ShabondiType {
```
Comment (chia7712) [inline ohara-shabondi/src/main/scala/oharastream/ohara/shabondi/ShabondiDefinitions.scala]: see oharastream.ohara.client.Enum

File: ohara-shabondi/src/main/scala/oharastream/ohara/shabondi/ShabondiDefinitions.scala [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+    .registerTo(basicDefinitionMap)
+
+  val IMAGE_NAME_DEFINITION = SettingDef.builder
+    .group(CORE_GROUP)
+    .key("imageName")
+    .orderInGroup(orderInGroup())
+    .displayName("Image name")
+    .documentation("The image name of this Shabondi running with")
+    .optional(IMAGE_NAME_DEFAULT)
+    .permission(SettingDef.Permission.READ_ONLY) // In manager, user cannot change the image name
```
Comment (chia7712) [inline ohara-shabondi/src/main/scala/oharastream/ohara/shabondi/ShabondiDefinitions.scala]: So we disable user to run different version of shabondi at runtime?

File: ohara-shabondi/src/main/scala/oharastream/ohara/shabondi/ShabondiDefinitions.scala [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+    .key("shabondi.brokers")
+    .orderInGroup(orderInGroup())
+    .required(Type.STRING)
+    .displayName("Broker list")
+    .documentation("The broker list of current workspace")
+    .internal
+    .build
+    .registerTo(basicDefinitionMap)
+
+  val ROUTES_DEFINITION = SettingDef.builder
```
Comment (chia7712) [inline ohara-shabondi/src/main/scala/oharastream/ohara/shabondi/ShabondiDefinitions.scala]: Are there extra routes used by shabondi? The routes to services (for example, broker) are configured by collie, and this filed is used to accept custom routes passed by users.

File: ohara-shabondi/src/main/scala/oharastream/ohara/shabondi/ShabondiDefinitions.scala [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+    .key("revision")
+    .orderInGroup(orderInGroup())
+    .displayName("Revision")
+    .documentation("Revision of stream")
+    .optional(VersionUtils.REVISION)
+    .permission(SettingDef.Permission.READ_ONLY)
+    .build
+    .registerTo(basicDefinitionMap)
+
+  val AUTHOR_DEFINITION: SettingDef = SettingDef.builder
```
Comment (chia7712) [inline ohara-shabondi/src/main/scala/oharastream/ohara/shabondi/ShabondiDefinitions.scala]: version, revision and author are moved to common group. Also oharastream.ohara.common.setting.WithDefinitions has many helper methods to generate those common definitions.

File: ohara-shabondi/src/main/scala/oharastream/ohara/shabondi/ShabondiDefinitions.scala (hunk 1 above)
Comment (vitojeng) [reply ohara-shabondi/src/main/scala/oharastream/ohara/shabondi/ShabondiDefinitions.scala]: Got it.

File: ohara-shabondi/src/main/scala/oharastream/ohara/shabondi/ShabondiDefinitions.scala (hunk 2 above)
Comment (vitojeng) [reply ohara-shabondi/src/main/scala/oharastream/ohara/shabondi/ShabondiDefinitions.scala]: The same permission setting with Stream.
I guess we would enable it in the future?

Comment (chia7712) [reply ohara-shabondi/src/main/scala/oharastream/ohara/shabondi/ShabondiDefinitions.scala]: > I guess we would enable it in the future?

absolutely. Could you leave a issue link for it?

File: ohara-shabondi/src/main/scala/oharastream/ohara/shabondi/ShabondiDefinitions.scala (hunk 3 above)
Comment (vitojeng) [reply ohara-shabondi/src/main/scala/oharastream/ohara/shabondi/ShabondiDefinitions.scala]: The extra route should not used by shabondi now. 
Will remove it.

File: ohara-shabondi/src/main/scala/oharastream/ohara/shabondi/ShabondiDefinitions.scala (hunk 2 above)
Comment (vitojeng) [reply ohara-shabondi/src/main/scala/oharastream/ohara/shabondi/ShabondiDefinitions.scala]: sure

File: ohara-shabondi/src/main/scala/oharastream/ohara/shabondi/ShabondiDefinitions.scala (hunk 4 above)
Comment (vitojeng) [reply ohara-shabondi/src/main/scala/oharastream/ohara/shabondi/ShabondiDefinitions.scala]: Got it, thanks for remind.

**Pull request merged**